scipy/optimize/_zeros_py.py

```diff
@@ -1388,3 +1388,278 @@ def toms748(f, a, b, args=(), k=1,
+    >>> k = np.array([[[1, 2]]])
+    >>> res = optimize._chandrupatla(f,x0=x0,x1=x1,
+                              args=(k,),
+                              maxiter=50)
+    >>> res.x
+    [[[ 1.000e+00  1.414e+00]
+         [ 1.000e+00  1.414e+00]]
+
+        [[ 1.000e+00  1.414e+00]
+         [ 1.000e+00  1.414e+00]]]
+
+    permission from Jason Sachs (@jason-s):
+        https://github.com/scipy/scipy/issues/7242#issuecomment-1314178133
+    """
+
+    # adapted from an earlier implementation of Jason Sachs'
+    # as written in https://www.embeddedrelated.com/showarticle/855.php
+    # which in turn is based on Chandrupatla's algorithm as described in
+    # Scherer https://books.google.com/books?id=cC-8BAAAQBAJ&pg=PA95
+
+    # Initialization
+    b = x0
+    a = c = x1
+    fb = f(b, *args)
+    fa = fc = f(a, *args)
+    t = 0.5
+    iterations = 0
+
+    a, b, fa, fb = np.broadcast_arrays(a, b, fa, fb)
+    intermediate_shape = a.shape
+    a, b, fa, fb = np.atleast_1d(a, b, fa, fb)
+    shape = a.shape
+
+    iterations = np.zeros_like(a)
+
+    # flag to check state of convergence
+    flag = np.full(shape, _EINPROGRESS)
+
+    # Initialize an array of False,
+    # determines whether we should do inverse quadratic interpolation
+    iqi = np.zeros(shape, dtype=bool)
+
+    funcalls = np.full(np.shape(flag), 0)
+    terminate = False
+
+    while iterations.any() < maxiter:
+        # use t to linearly interpolate between a and b,
+        # and evaluate this function as our newest estimate xt
+        xt = a + t * (b - a)
+        ft = f(xt, *args)
+        funcalls += 1
+
+        # update our history of the last few points so that
+        # - a is the newest estimate (we're going to update it from xt)
+        # - c and b get the preceding two estimates
+        # - a and b maintain opposite signs for f(a) and f(b)
+        samesign = np.sign(ft) == np.sign(fa)
+        c = np.choose(samesign, [b, a])
+        b = np.choose(samesign, [a, b])
+        fc = np.choose(samesign, [fb, fa])
+        fb = np.choose(samesign, [fa, fb])
+        a = xt
+        fa = ft
+
+        # set r so that f(r) is the minimum magnitude of f(a) and f(b)
+        fa_is_smaller = np.abs(fa) < np.abs(fb)
+        r = np.choose(fa_is_smaller, [b, a])
+        fm = np.choose(fa_is_smaller, [fb, fa])
+
+        tol = 2 * xatol * np.abs(r) + xrtol
+        tlim = tol / np.abs(b - c)
+        terminate = np.logical_or(terminate,
+                                  np.logical_or(fm == 0, tlim > 0.5))
+
+        # check convergence here, update flag if so
+        if np.all(terminate):
+            flag.fill(_ECONVERGED)
+            break
+
+        iterations += 1 - terminate
+
+        # Figure out values xi and phi
+        # to determine which method we should use next
+        xi = (a - b) / (c - b)
+        phi = (fa - fb) / (fc - fb)
+        iqi = np.logical_and(phi**2 < xi, (1 - phi) ** 2 < 1 - xi)
+
+        if iqi.all():
```

Comment:
I haven't re-checked, but this doesn't look right to me. Why would _all_ `iqi` have to be true in order to do something that uses `iqi` as a boolean index?

Reply:
Hmm good question, this was my attempt at integrating the scalar and array cases. Do you think this should be `if iqi.any()` or something else entirely?

Reply:
I don't think you really need an `if` statement for this sort of thing. `where`, `lazywhere`, or just logical indexing does the trick. Here is my code, for example, from the first vectorized version in https://github.com/scipy/scipy/issues/7242#issuecomment-1363383616 :

```python3
        # Flowchart 4
        # Equation 1
        xi1 = (x1 - x2) / (x3 - x2)
        phi1 = (f1 - f2) / (f3 - f2)
        alpha = (x3 - x1) / (x2 - x1)
        t = f1/(f1 - f2) * f3/(f3 - f2) - alpha * f1/(f3-f1) * f2/(f2-f3)
        j = ((1 - np.sqrt(1 - xi1)) >= phi1) | (phi1 >= np.sqrt(xi1))
        t[j] = 0.5
```

Reply:
Great, thank you - I'll try that and see if I can work it in.

Reply:
Sorry for the delay, still working on this